src/compute/src/sink/correction_v2.rs

```diff
@@ -0,0 +1,1109 @@
+    fn last(&self) -> (&D, Timestamp, Diff) {
+        self.index(self.len() - 1)
+    }
+
+    /// Push an update onto the chunk.
+    fn push<DT: Borrow<D>>(&mut self, update: (DT, Timestamp, Diff)) {
+        let (d, t, r) = update;
+        self.data.copy_destructured(d.borrow(), &t, &r);
+
+        self.invalidate_cached_size();
+    }
+
+    /// Return the index of the first update at a time greater than `time`, or `None` if no such
+    /// update exists.
+    fn find_time_greater_than(&self, time: Timestamp) -> Option<usize> {
+        if self.last().1 <= time {
+            return None;
+        }
+
+        let mut lower = 0;
+        let mut upper = self.len();
+        while lower < upper {
+            let idx = (lower + upper) / 2;
+            if self.index(idx).1 > time {
+                upper = idx;
+            } else {
+                lower = idx + 1;
+            }
+        }
+
+        Some(lower)
+    }
+
+    /// Return the size of the chunk, for use in metrics.
+    fn get_size(&mut self) -> LengthAndCapacity {
+        if self.cached_size.is_none() {
+            let length = Container::len(&self.data);
+            let mut capacity = 0;
+            self.data.heap_size(|_, cap| capacity += cap);
+            self.cached_size = Some(LengthAndCapacity { length, capacity });
+        }
+
+        self.cached_size.unwrap()
+    }
+
+    /// Invalidate the cached chunk size.
+    ///
+    /// This method must be called every time the size of the chunk changed.
+    fn invalidate_cached_size(&mut self) {
+        self.cached_size = None;
+    }
+}
+
+/// Sort and consolidate the given list of updates.
+///
+/// This function is the same as [`differential_dataflow::consolidation::consolidate_updates`],
+/// except that it sorts updates by (time, data) instead of (data, time).
+fn consolidate<D: Data>(updates: &mut Vec<(D, Timestamp, Diff)>) {
+    if updates.len() <= 1 {
+        return;
+    }
+
+    let diff = |update: &(_, _, Diff)| update.2;
+
+    updates.sort_unstable_by(|(d1, t1, _), (d2, t2, _)| (t1, d1).cmp(&(t2, d2)));
+
+    let mut offset = 0;
+    let mut accum = diff(&updates[0]);
+
+    for idx in 1..updates.len() {
+        let this = &updates[idx];
+        let prev = &updates[idx - 1];
+        if this.0 == prev.0 && this.1 == prev.1 {
+            accum += diff(&updates[idx]);
+        } else {
+            if accum != 0 {
+                updates.swap(offset, idx - 1);
+                updates[offset].2 = accum;
+                offset += 1;
+            }
+            accum = diff(&updates[idx]);
+        }
+    }
+
+    if accum != 0 {
+        let len = updates.len();
+        updates.swap(offset, len - 1);
+        updates[offset].2 = accum;
+        offset += 1;
+    }
+
+    updates.truncate(offset);
+}
+
+/// Merge the given chains, advancing times by the given `since` in the process.
+fn merge_chains<D: Data>(chains: Vec<Chain<D>>, since: &Antichain<Timestamp>) -> Chain<D> {
```

Comment:
```suggestion
fn merge_chains<D: Data>(chains: impl Iterator<Chain<D>>, since: &Antichain<Timestamp>) -> Chain<D> {
```

Reply:
Changed to `impl IntoIterator` because that's more convenient to use by callers.